Introduce experimental accessible colors in markdown rendering

andyfeller opened the pull request:
Relates github/cli#830

These changes relate to a new experimental feature for rendering markdown in GitHub CLI and CLI extensions using accessible colors.

Out of the box, `markdown` package continues to render GFM using existing `charmbracelet/glamour` light and dark styles in 8-bit color.

By setting `accessible_colors` configuration setting or `GH_ACCESSIBLE_COLORS` environment variable, `markdown` package will begin using ANSI 4-bit colors that most modern terminals allow users to customize.  This will allow users with low vision and color blindness to tailor their experiences more effectively.

## Go version change

This pull request bumps the minimum Go version from 1.21 to 1.23 due to transitive dependencies from experimental Go module:

- https://github.com/charmbracelet/glamour/pull/396
- https://github.com/charmbracelet/glamour/pull/397

## Demo

The following accessible colors have been aligned for consistency with [Primer Design System](https://primer.style/) within the limited palette of 4-bit colors.

### Headers

**Current default**

<img width="1512" alt="headers, current release, dark appearance" src="https://github.com/user-attachments/assets/fd02f752-a9c4-4a65-b554-52e101ed6a16" />

<img width="1512" alt="headers, current release, light apperance" src="https://github.com/user-attachments/assets/1135363b-59c7-4f3a-8e43-315ffaef6609" />

**Accessible**

<img width="1512" alt="headers, accessible-colors, dark appearance" src="https://github.com/user-attachments/assets/143dd4ae-322e-4a96-bd8d-c59d48b51863" />

<img width="1512" alt="headers, accessible-colors, light apperance" src="https://github.com/user-attachments/assets/8908f27c-0ff2-4d1d-b38f-bcd0abd3fe25" />

**New default**

<img width="1512" alt="headers, new default, dark appearance" src="https://github.com/user-attachments/assets/a34ad629-2532-4483-821b-3289776f5e45" />

<img width="1512" alt="headers, new default, light appearance" src="https://github.com/user-attachments/assets/3aff1e6e-8517-4534-b357-06b422c7429d" />


### Horizontal Rule

**Current default**

<img width="1512" alt="horizontal rule, current release, dark appearance" src="https://github.com/user-attachments/assets/2082fb6e-b9ad-4fc7-895d-b87686bb9e45" />

<img width="1512" alt="horizontal rule, current release, light appearance" src="https://github.com/user-attachments/assets/e07c59cf-7732-46ad-88fd-c354b9f70e24" />

**Accessible**

<img width="1512" alt="horizontal rule, accessible-colors, dark appearance" src="https://github.com/user-attachments/assets/e6cf9233-8732-4a53-a6b5-d10d6f7d7548" />

<img width="1512" alt="horizontal rule, accessible-colors, light appearance" src="https://github.com/user-attachments/assets/c87db5e2-48f3-4a5a-8057-0ae639beaf10" />

**New default**

<img width="1512" alt="horizontal rule, new default, dark appearance" src="https://github.com/user-attachments/assets/1d12e9d9-a4e5-455f-9093-4547d3b6efa9" />

<img width="1512" alt="horizontal rule, new default, light appearance" src="https://github.com/user-attachments/assets/f686636d-ce15-4164-8464-1ac385b2ea2a" />


### Images

**Current default**

<img width="1512" alt="image, current release, dark appearance" src="https://github.com/user-attachments/assets/beb0df04-eaa8-466b-8370-fd02b5b09bde" />

<img width="1512" alt="image, current release, light appearance" src="https://github.com/user-attachments/assets/d2a875f9-df95-4c59-a85e-3d954e250822" />

**Accessible**

<img width="1512" alt="image, accessible-colors, dark appearance" src="https://github.com/user-attachments/assets/feff5ad6-4a85-47bc-8f5f-1dc3e165262e" />

<img width="1512" alt="image, accessible-colors, light appearance" src="https://github.com/user-attachments/assets/71821ac8-23b7-4872-a414-89eb36a71558" />

**New default**

<img width="1512" alt="image, new default, dark appearance" src="https://github.com/user-attachments/assets/6d2b3bdd-8b5f-44c2-ada7-583aca7547e6" />

<img width="1512" alt="image, new default, light appearance" src="https://github.com/user-attachments/assets/8306a2f0-6e69-47c3-9bbe-4ba5b3bb26f9" />


### Links

**Current default**

<img width="1512" alt="link, current release, dark appearance" src="https://github.com/user-attachments/assets/3cdb0c25-6437-4b47-9cb4-99b379f07af1" />

<img width="1512" alt="link, current release, light appearance" src="https://github.com/user-attachments/assets/c3545d71-1aed-426e-8e12-bd160c7d4657" />

**Accessible**

<img width="1512" alt="link, accessible-colors, dark appearance" src="https://github.com/user-attachments/assets/d859a5fd-4703-4835-ad01-8f5a64348190" />

<img width="1512" alt="link, accessible-colors, light appearance" src="https://github.com/user-attachments/assets/3c01d937-19d9-4d23-8dc1-bca65b4a03b2" />

**New default**

<img width="1512" alt="link, new default, dark appearance" src="https://github.com/user-attachments/assets/7e5b6856-36ef-452b-b7af-6638fe3310ce" />

<img width="1512" alt="link, new default, light appearance" src="https://github.com/user-attachments/assets/9cd3ff3b-e0b7-4767-9449-2da0ce6c2fce" />


### Code

**Current default**

<img width="1512" alt="code, current release, dark appearance" src="https://github.com/user-attachments/assets/eba8f35d-58d0-48a1-bfe3-0fefd56a3e25" />

<img width="1512" alt="code, current release, light appearance" src="https://github.com/user-attachments/assets/e0dd4420-3c20-4993-9cb2-82990dc57afe" />

**Accessible**

<img width="1512" alt="code, accessible-colors, dark appearance" src="https://github.com/user-attachments/assets/d7864d5d-451e-429d-aa4d-310286aaa24c" />

<img width="1512" alt="code, accessible-colors, light appearance" src="https://github.com/user-attachments/assets/8dde51ea-d07c-4f87-9487-76a06d9c0f83" />

**New default**

<img width="1512" alt="code, new default, dark appearance" src="https://github.com/user-attachments/assets/bdf63daf-adcb-4bae-a6f8-8a5345c7d068" />

<img width="1512" alt="code, new default, light appearance" src="https://github.com/user-attachments/assets/792e411b-0f7f-45ce-b5e5-fef706ef5633" />


### Quotes

**Current default**

<img width="1512" alt="quotes, current release, dark appearance" src="https://github.com/user-attachments/assets/2adb0329-9852-4ae4-a9d4-4b64ec5fc0e0" />

<img width="1512" alt="quotes, current release, light appearance" src="https://github.com/user-attachments/assets/9d68bdc5-f75d-4e12-8d04-d368cd312c2c" />

**Accessible**

<img width="1512" alt="quotes, accessible-colors, dark appearance" src="https://github.com/user-attachments/assets/ed100a5e-abcf-4df8-a691-4e15da7f8941" />

<img width="1512" alt="quotes, accessible-colors, light appearance" src="https://github.com/user-attachments/assets/3bdffc9e-5412-489c-a9d9-217c36a0a668" />

**New default**

<img width="1512" alt="quotes, new default, dark appearance" src="https://github.com/user-attachments/assets/c39a60b9-ad55-4186-a44d-fbe24dd0b3a4" />

<img width="1512" alt="quotes, new default, light appearance" src="https://github.com/user-attachments/assets/ea9ff9cc-adc6-49ad-8d99-d9383cece902" />


### Tables

**Current default**

<img width="1512" alt="table, current release, dark appearance" src="https://github.com/user-attachments/assets/79fa72cb-1c9e-432e-942f-27517980ab68" />

<img width="1512" alt="table, current release, light appearance" src="https://github.com/user-attachments/assets/0c88b474-212d-4764-b655-55cba7c7ac62" />

**Accessible**

<img width="1512" alt="table, accessible-colors, dark appearance" src="https://github.com/user-attachments/assets/b429b0a7-722e-405c-bbd5-990552cee333" />

<img width="1512" alt="table, accessible-colors, light appearance" src="https://github.com/user-attachments/assets/8644a80a-a171-437c-a7f4-d991950a72e5" />

**New default**

<img width="1512" alt="table, new default, dark appearance" src="https://github.com/user-attachments/assets/16932cfd-3209-470b-9790-3f791cf7bb68" />

<img width="1512" alt="table, new default, light appearance" src="https://github.com/user-attachments/assets/e679c176-9972-4f4f-8b63-6374b5131aab" />

jtmcg committed:
Move duplicated stubConfig function to a new internal/testutils dir

jtmcg committed:
Add new accessibility package and IsEnabled() function

jtmcg committed:
Address PR comments

jtmcg committed:
Change GH_ACCESSIBLE env var to ACCESSIBLE

jtmcg committed:
Merge pull request #180 from cli/jtmcg/accessibility-isEnabled

Add accessibility package and isEnabled function

jtmcg committed:
Add accessible themes to markdown rendering

jtmcg committed:
Add AnsiColor enum

jtmcg committed:
Add accessible styles

jtmcg committed:
Cleaned up naming and comments for better understandability

jtmcg committed:
Upgrade glamour version for access to styles package

jtmcg committed:
Fix typo in env var name

jtmcg committed:
Change link stylings so link text is correct color

jtmcg committed:
go mod tidy

jtmcg committed:
Update test to check correct config

jtmcg committed:
Fix typo

jtmcg committed:
Merge pull request #182 from cli/jtmcg/add-accessible-themes

Add accessible themes

andyfeller committed:
Ensure codeblocks use accessible colors

This commit leverages `cli` fork of `charmbracelet/glamour` with enhancement for configuring `chroma` formatter.

It includes simple tests that enabling `gh` accessible features causes codeblocks to be downsampled to base 16 ANSI colors.

andyfeller committed:
v2 iteration on testing codeblocks

This commit is v2 approach to testing for accessible colors and codeblock rendering using an in-house approach to identifying ANSI escape sequences and analyzing them for color depth.

After talking with @williammartin, this is likely going to be refactored a third time leveraging a module to help with parsing escape sequences from text.

andyfeller committed:
Simplify markdown codeblock tests, glamour helper

This commit is focused on PR feedback around codeblock testing and simplifying related code:

1. Use of new `WithOptions(...TermRendererOption) TermRendererOption` to clean up `WithTheme()`

   The `glamour` TermRendererOption pattern has a limitation that users cannot compose multiple options without duplicating code or building one-off anonymous functions.

   However, this commit takes advantage of an enhancement in https://github.com/cli/glamour/pull/3 that allows users to leverage a helper to avoid building one-offs.

1. Use of new `leaanthony/go-ansi-parser` dependency for parsing ANSI escape sequences and display attributes

   In v1 of `markdown_test.go`, the codeblock tests were very simple, testing the result of output of markdown rendering against a string of ANSI escape sequences. The concern raised is that this was testing the result rather than behavior wanted.

   In v2 of `markdown_test.go`, the codeblock tests were refactored to use regex to extract and analyze ANSI escape sequences and display attributes. The concern raised was that this was a lot of logic to build and maintain and might benefit from a dependency that could do it.

   In v3 of `markdown_test.go`, a combination of v1 and v2 approaches for 1) testing that theme appropriate colors are used and 2) testing that ensures accessible display options are used when accessible experience is enabled

andyfeller committed:
PR feedback

After discussing the changing nature of the tests with @jtmcg, this commit renames the test functions to make it apparent their purposes: testing we get the colors we want separate from checking for accessibility concerns.

In addition, there were a number of other improvements to be more idiomatic.

andyfeller committed:
Merge pull request #183 from cli/andyfeller/accessible-colors-chroma

Ensure codeblocks use accessible colors

andyfeller committed:
Realign accessible color logic around new env var

1. Introduce constant for `gh config` setting for accessible colors
2. Align existing constant for env var to match
3. Rename exported functions to distinguish use specifically for color
4. Minor enhancements to comments and tests

andyfeller committed:
Reorganize and communicate experimental status

Following Go `x` convention seen across various projects, the former `accessibility` package has been relocated to communicate its experimental status more clearly to extension authors.

It has also been renamed to `color` to align with adopting Primer Style color role functionality.

andyfeller committed:
Fix bugs from package renaming

andyfeller committed:
Merge pull request #185 from cli/andyfeller/gh_accessible_colors

Realign accessible color logic around new env var

andyfeller committed:
Introduce experimental accessibile markdown colors

Relates github/cli#830

andyfeller committed:
Workflows should use go.mod for Go version

Due to errors in PR, this commit updates GitHub Actions workflows to use a newer version of actions/setup-go capable of reading Go version from go.mod.

andyfeller committed:
Reorder workflows to checkout first

andyfeller commented:
In addressing the lint errors, I've made minor changes to GitHub Actions workflows to pull Go version from `go.mod`, which requires a newer version of `actions/setup-go`.

Separately while troubleshooting lint job, I notice that the action used for Go linting is 3 major version behind; [`v3.4.0`](https://github.com/golangci/golangci-lint-action/releases/tag/v3.4.0) => [`v6.5.0`](https://github.com/golangci/golangci-lint-action/releases/tag/v6.5.0).

For changes, see [diff](https://github.com/golangci/golangci-lint-action/compare/v3.4.0...v6.5.0)

andyfeller committed:
Merge branch 'trunk' into accessible-colors

andyfeller committed:
Bump golangci-lint action for failing lint job

Iterate on resolving lint job errors due to Go 1.23 version upgrade.

This should resolve most of the problems with 1 potential fix that I'll wait to confirm in the next run.

andyfeller committed:
Address golangci-lint problem

williammartin commented on .github/workflows/lint.yml:
```diff
@@ -7,20 +7,21 @@ jobs:
         with:
-          version: v1.54
+          version: v1.64
+          problem-matchers: true
```
**Question**

I read the [doc](https://github.com/golangci/golangci-lint-action?tab=readme-ov-file#problem-matchers) but I don't understand what this is doing. Help?

andyfeller replied:
### Background

One of the most esoteric features of GitHub Actions is the ability to use annotations, which CI can draw attention to problems or other notable aspects during workflow run.  Using one of the following [Workflow commands](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions), you are describing where GitHub Actions should display the annotation in the GitHub UI, which pops up as a little box to draw the viewers attention:

- [Setting a notice message](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-a-notice-message)

  ```
  ::notice file={name},line={line},endLine={endLine},title={title}::{message}
  ```

- [Setting a warning message](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-a-warning-message)

  ```
  ::notice file={name},line={line},endLine={endLine},title={title}::{message}
  ```
  
- [Setting an error message](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-an-error-message)

  ```
  ::error file={name},line={line},endLine={endLine},title={title}::{message}
  ```

However, this is a very specific format which people might not like or you might use tools that you don't control but want to bubble up certain problems.  This is where a [**problem matcher**](https://github.com/actions/runner/blob/0ce29d09c64461774c26cd23bc4610d3e94b1edd/docs/adrs/0276-problem-matchers.md#L4) is necessary to identify and translate output from GitHub Actions workflow run into annotations.  I've only found docs on this in the `actions/runner` ADRs; nothing in GitHub Docs.

### `actions/setup-go` problem matcher and `problem-matchers: true`

Now back to this change, `actions/setup-go` sets up a custom problem matcher, however I'm unsure the full story behind the format due to the lack of docs and commentary in https://github.com/actions/setup-go/pull/2:

- [`src/main.ts`](https://github.com/actions/setup-go/blob/5a083d0e9a84784eb32078397cf5459adecb4c40/src/main.ts#L81-L83) sets up a custom problem matcher

  ```typescript
      // add problem matchers
      const matchersPath = path.join(__dirname, '../..', 'matchers.json');
      core.info(`##[add-matcher]${matchersPath}`);
  ```


- [`matchers.json`](https://github.com/actions/setup-go/blob/5a083d0e9a84784eb32078397cf5459adecb4c40/matchers.json) seems like it matches a generic format with nothing like the log-like priority of workflow commands

  ```json
  {
      "problemMatcher": [
          {
              "owner": "go",
              "pattern": [
                  {
                      "regexp": "^\\s*(.+\\.go):(?:(\\d+):(\\d+):)? (.*)",
                      "file": 1,
                      "line": 2,
                      "column": 3,
                      "message": 4
                  }
              ]
          }
      ]
  }
  ```

As for `golangci/golangci-lint-action`, it appears to be designed to conditionally setup problem matchers if `problem-matchers: true` and it has a slightly different setup than `actions/setup-go`:

- [`action.yml`](https://github.com/golangci/golangci-lint-action/blob/b871b4f9de8b829c8d43b0e3bcdf9a8a3f141a3b/action.yml#L45-L48) 

  ```yaml
    problem-matchers:
      description: "Force the usage of the embedded problem matchers"
      default: 'false'
      required: false
  ```

- [`src/run.ts`](https://github.com/golangci/golangci-lint-action/blob/b871b4f9de8b829c8d43b0e3bcdf9a8a3f141a3b/src/run.ts#L68-L77)

  ```typescript
    const problemMatchers = core.getBooleanInput(`problem-matchers`)
  
    if (problemMatchers) {
      const matchersPath = path.join(__dirname, "../..", "problem-matchers.json")
      if (fs.existsSync(matchersPath)) {
        // Adds problem matchers.
        // https://github.com/actions/setup-go/blob/cdcb36043654635271a94b9a6d1392de5bb323a7/src/main.ts#L81-L83
        core.info(`##[add-matcher]${matchersPath}`)
      }
    }
  ```

- [`problem-matchers.json`](https://github.com/golangci/golangci-lint-action/blob/b871b4f9de8b829c8d43b0e3bcdf9a8a3f141a3b/problem-matchers.json)

  ```json
  {
    "problemMatcher": [
      {
        "owner": "golangci-lint-colored-line-number",
        "severity": "error",
        "pattern": [
          {
            "regexp": "^([^:]+):(\\d+):(?:(\\d+):)?\\s+(.+ \\(.+\\))$",
            "file": 1,
            "line": 2,
            "column": 3,
            "message": 4
          }
        ]
      }
    ]
  }
  ```

### Which should we use?

1. `actions/setup-go`

   ```json
   "regexp": "^\\s*(.+\\.go):(?:(\\d+):(\\d+):)? (.*)"
   ```

1. `golangci/golangci-lint-action` with an `error` severity

   ```json
   "regexp": "^([^:]+):(\\d+):(?:(\\d+):)?\\s+(.+ \\(.+\\))$"
   ```

Between the 2, I would assume the problem matcher setup by `golangci` is more appropriate for its use than the generic, non-severity problem matcher from `actions/setup-go`.

![the end](https://github.com/user-attachments/assets/bef3eaae-09ad-427f-b375-cc9c466ecff2)

williammartin replied:
wat

williammartin replied:
Thanks for the explanation. I guess at some point it would be good to see a comparison of the two visually.

andyfeller replied:
Additional documentation on problem matchers:
https://github.com/actions/toolkit/blob/7ae578ddd11b9e33db1daaad2ffb39e1ffeada3f/docs/problem-matchers.md

williammartin commented on .github/workflows/lint.yml:
```diff
@@ -7,20 +7,21 @@ jobs:
       - name: Checkout repository
         uses: actions/checkout@v4
 
+      - name: Set up Go
```
**Question**

Why did you reorder the checkout and setup? Same for `lint.yml`

andyfeller replied:
In order for `actions/setup-go@v5` to use the `go.mod` for determining the Go toolchain version.  These workflows having a hardcoded Go version makes situations where the version changes a little more painful than needed.

williammartin replied:
Got it, thanks!

williammartin commented on internal/yamlmap/yaml_map.go:
```diff
@@ -72,7 +72,7 @@ func (m *Map) AddEntry(key string, value *Map) {
 
 func (m *Map) Empty() bool {
-	return m.Content == nil || len(m.Content) == 0
+	return len(m.Content) == 0
```
**Question**

Can you explain this change please? I'm guessing it was linted because the `nil` slice has `len == 0` but just want to check!

andyfeller replied:
Great question!  This was the only error / warning raised by `golangci-lint`:

```shell
$ golangci-lint run --out-format=colored-line-number
internal/yamlmap/yaml_map.go:75:9: S1009: should omit nil check; len() for nil slices is defined as zero (gosimple)
	return m.Content == nil || len(m.Content) == 0
	       ^
```

williammartin replied:
Cool, thanks.

williammartin commented on internal/testutils/config_stub.go:
```diff
@@ -0,0 +1,21 @@
+// StubConfig replaces the config.Read function with a function that returns a config object
+// created from the given config string. It also sets up a cleanup function that restores the
+// original config.Read function.
+func StubConfig(t *testing.T, cfgStr string) {
```
**Food For Thought**

No concerns merging this as is because it's easy to change later, but maybe this should be `config.Stub` where this lives in `package config_test` would be better than a `testutils` package.

andyfeller replied:
I'm down for this, but would probably like to sync over a weird behavior around importing the package I'm seeing while acting on this. 🤔

williammartin replied:
On second thoughts it might not be possible to import from a different package for `_test` tags. Sorry for the wild goose chase.

williammartin commented on pkg/markdown/accessibility.go:
```diff
@@ -0,0 +1,106 @@
+	return strconv.Itoa(int(a))
+}
+
+func (a ANSIColorCode) StrPtr() *string {
```
**Question**

Did you take this pattern from somewhere?

andyfeller replied:
Yeah, I believe it was based on 2 ideas:

- [`charmbracelet/glamour` default styles logic](https://github.com/charmbracelet/glamour/blob/4c040b7fd5c023154de93d5c0f789111ea06c82c/styles/styles.go#L135-L343):

  ```go
	  // DarkStyleConfig is the default dark style.
	  DarkStyleConfig = ansi.StyleConfig{
		  Document: ansi.StyleBlock{
			  StylePrimitive: ansi.StylePrimitive{
				  BlockPrefix: "\n",
				  BlockSuffix: "\n",
				  Color:       stringPtr("252"),
			  },
			  Margin: uintPtr(defaultMargin),
		  },
		  BlockQuote: ansi.StyleBlock{
			  StylePrimitive: ansi.StylePrimitive{},
			  Indent:         uintPtr(1),
			  IndentToken:    stringPtr("│ "),
		  },
  ```
  
  which uses a handful of [helper functions](https://github.com/charmbracelet/glamour/blob/4c040b7fd5c023154de93d5c0f789111ea06c82c/styles/styles.go#L678-L680):
  
  ```go
  func boolPtr(b bool) *bool       { return &b }
  func stringPtr(s string) *string { return &s }
  func uintPtr(u uint) *uint       { return &u }
  ```

- [`color.go` in `cli/cli` and `ansi.ColorFunc()`](https://github.com/cli/cli/blob/af4acb380136fd106b38cc0ab404ff975bca9795/pkg/iostreams/color.go#L15-L33)

  This was more inspirational about having a type where information about color could then be leveraged elsewhere rather than the unexported private functions above:

  ```go
  var (
	  magenta            = ansi.ColorFunc("magenta")
	  cyan               = ansi.ColorFunc("cyan")
	  red                = ansi.ColorFunc("red")
	  yellow             = ansi.ColorFunc("yellow")
	  blue               = ansi.ColorFunc("blue")
	  green              = ansi.ColorFunc("green")
	  gray               = ansi.ColorFunc("black+h")
	  lightGrayUnderline = ansi.ColorFunc("white+du")
	  bold               = ansi.ColorFunc("default+b")
	  cyanBold           = ansi.ColorFunc("cyan+b")
	  greenBold          = ansi.ColorFunc("green+b")
	  highlightStart     = ansi.ColorCode(highlightStyle)
	  highlight          = ansi.ColorFunc(highlightStyle)
  
	  gray256 = func(t string) string {
		  return fmt.Sprintf("\x1b[%d;5;%dm%s\x1b[m", 38, 242, t)
	  }
  )
  ```
```

williammartin replied:
Alright, makes sense to me.

Maybe it doesn't need to be exported?

andyfeller replied:
If we're moving this under `x`, don't we have to export it?

williammartin replied:
It kind of looked like `StrPtr` was only used in this package. Am I missing something?

andyfeller replied:
No you aren't 🚀 let's unexport it.

williammartin commented on pkg/markdown/accessibility.go:
```diff
@@ -0,0 +1,106 @@
+package markdown
```
**Question**

Should this also live under `x` for the moment?

andyfeller replied:
Great question!  I can get behind that especially as we don't want anyone thinking its stable, yeah.

andyfeller replied:
Related question: any preference on the alias used when importing `x` packages?

I know there are broader Go conventions on the topic, but I don't think I've encountered another project that ran into this situation to have a strong opinion and thus went with:

```go
// Package markdown facilitates rendering markdown in the terminal.
package markdown

import (
	"os"
	"strings"

	"github.com/charmbracelet/glamour"
	xcolor "github.com/cli/go-gh/v2/pkg/x/color"
	xmarkdown "github.com/cli/go-gh/v2/pkg/x/markdown"
)
```

williammartin replied:
No opinions.

williammartin commented on pkg/x/color/accessibility.go:
```diff
@@ -0,0 +1,40 @@
+}
+
+func isEnvVarEnabled(envVar string) bool {
+	return envVar != "" && envVar != "0" && envVar != "false"
```
**Food For Thought**

No specific request but this function looks a little sus in that the only place it is used is in this file and repeats the string check that wraps it on line 26.

I think I'd probably inline on 26 like if you wanted to keep the clarity:

```go
if envVar != "" {
   isEnVarEnabled := envVar != "0" && envVar != "false"
   return isEnvVarEnabled
}
```

andyfeller replied:
Totally fair call out!

I was hoping we could tie into existing logic as this work evolved as this shouldn't be a new problem, but I suppose we don't really:

https://github.com/cli/go-gh/blob/7a6e5543a7f13a372dfafa88fac99981496b4991/pkg/api/http_client.go#L80-L88

andyfeller replied:
We probably also want to add `no`, too

andyfeller commented:
> [!NOTE]
> In upping the Go version, there is a discrepancy between automation in the default branch which takes 1.21 into account whereas this branch bumps it up. 🤷 
>
> Meaning there will be some follow up work to the repository ruleset / branch protection rules afterwards.

williammartin reviewed (commented):
@andyfeller I got a bit distracted with some other stuff so I didn't get into the good stuff yet in this PR review sorry. Mostly just cursory stuff!

williammartin commented on pkg/x/color/accessibility_test.go:
```diff
@@ -0,0 +1,92 @@
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Setenv("GH_ACCESSIBLE_COLORS", tt.envVarValue)
```
**Food For Thought**

In practice I think this will work but I'm not sure it's totally correct in line with the descriptions e.g. the first test says:

```
When the accessibility configuration and env var are both unset, it should return false
```

But setting an env var to an empty string is not the same as it being unset. You can confirm this with:

```go
			_, set := os.LookupEnv("GH_ACCESSIBLE_COLORS")
			fmt.Println(set)
```

williammartin replied:
Actually if you look at this test:

```go
		{
			name:        "When the accessibility configuration is set to disabled and the env var is unset, it should return false",
			envVarValue: "",
			cfgStr:      accessibilityDisabledConfig(),
			wantOut:     false,
		},
```

You might expect that changing the implementation of `IsAccessibleColorsEnabled` with a refactor to a true "set" or "unset" check would be safe:

```go
	envVar, set := os.LookupEnv(AccessibleColorsEnv)
	if set {
		return isEnvVarEnabled(envVar)
	}
```

But it's not because `""` is considered set. This refactor actually results in the config **never** being checked in the tests.

For me, this is enough to **Request Changes** to make it clear what an empty string vs unset should do.

andyfeller replied:
Ooh good catch!

williammartin commented on pkg/x/color/accessibility_test.go:
```diff
@@ -0,0 +1,92 @@
+func accessibilityEnabledConfig() string {
+	return heredoc.Docf(`
+		%s: enabled
+	`, AccessibleColorsSetting)
```
**Nit**

I understand your desire to DRY out the use of this config but I have mixed feelings about it. The thing is, the tests are supposed to act as guardrails but changing the value of `AccessibleColorsSetting` would break users but wouldn't fail the tests.

I think it's generally better to duplicate in tests for this reason.

williammartin replied:
In fact you avoided this trap for the envvar by using the string directly in the test.

andyfeller replied:
Makes sense!

williammartin commented on pkg/markdown/accessibility_test.go:
```diff
@@ -0,0 +1,118 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestANSIColorCodeString(t *testing.T) {
```
**Question**

Maybe this is pedantic but it feels like we should test all the colors here. I swapped the order of some of them and no tests failed.

Maybe it's unnecessary, who's really going to swap the order? Maybe a comment would be fine. I don't think it's obvious to uninformed readers that the order of the enumeration is important.

An alternative would be to use a map to `int`.

What do you think?

andyfeller replied:
There isn't a comment in https://github.com/cli/go-gh/pull/182 but I recall at the time bringing up some past experiences relying upon enum and iota in other codebases.  Now this has been brought up completely out of the blue, I'm more confident maybe this changes.

andyfeller replied:
So after a bunch of back and forth, here are my thoughts:

1. The enumeration is the only way we can really do compile time type checking, so will leave it for now
2. I think ANSI color code is a little misleading _(technically ANSI colors = 3-bit)_
3. I don't know if we need separate `String` and `StrPtr` functions
4. I think renaming this to explain its purpose might be easier for people to understand; basically these are glamour style color code constants
5. I agree we should test all the colors

I pushed some changes up and happy to talk through them.

williammartin commented on pkg/markdown/markdown.go:
```diff
@@ -33,9 +34,16 @@ func WithWrap(w int) glamour.TermRendererOption {
 // If the environment variable GLAMOUR_STYLE is set, it will take precedence over the provided theme.
 func WithTheme(theme string) glamour.TermRendererOption {
 	style := os.Getenv("GLAMOUR_STYLE")
+	accessible := color.IsAccessibleColorsEnabled()
```
**Nit**

If this is only used inside the `light`, `dark` case then maybe just inline it in the `if` there?

It would probably be a totally negligible performance improvement but generally I value the scoping being tight.

williammartin commented on pkg/markdown/markdown_test.go:
```diff
@@ -1,38 +1,282 @@
+// Test_RenderAccessible tests rendered markdown for accessibility concerns such as color mode / depth and other display attributes.
+// It works by parsing the rendered markdown for ANSI escape sequences and checking their display attributes.
+// Test scenarios allow multiple color mode / depths because `ansi.Parse()` considers `\x1b[0m` sequence as part of `ansi.Default`.
+func Test_RenderAccessible(t *testing.T) {
```
**Praise**

This test is sick.

williammartin commented on pkg/markdown/markdown_test.go:
**Praise**

Great comment.

williammartin commented on pkg/markdown/markdown_test.go:
**Nit**

Same issue as DRYing the config setting mentioned in another comment.

williammartin commented on pkg/markdown/markdown_test.go:
**Question**

Is there a reason we're not using `require.NoError(err)` here? Suspect not.

andyfeller replied:
I'm unsure and comfortable changing it.

What are situations you would expect `t.Fatal(...)` being used?

williammartin replied:
None, if you're using `testify` across the codebase. If you're not using an assertion library, then I would expect to see it.

williammartin reviewed (changes requested):
Lots of good stuff in here, don't think there's too much work to resolve my comments.

I haven't yet pulled this into `cli/cli` to use it.

andyfeller committed:
Rework accessible colors env var handling, feedback

This commit is focused a handful of PR feedback:

1. Minor DRYing up with inlining logic
2. Update tests to hardcode env var and config setting rather than reuse constant names
3. Rework `GH_ACCESSIBLE_COLORS` evaluation (see below)

Previously, the env var logic had 2 concerns:

1. It couldn't distinguish between the env var being missing versus being empty
1. It didn't include `no` handling similar to `GH_DEBUG`

Now, `IsAccessibleColorsEnabled` will short circuit env var handling if unset.

andyfeller committed:
Relocate experimental packages, cleanup

Major change with this commit is relocating experimental logic into the `x`
package to communicate clearly expectations to users.  Additionally, the
package in-line docs have been updated to state experimental nature more
clearly.

There are a few other odd cleanups:

- test assertion improvement
- better in-line comments

andyfeller committed:
Unexport experimental markdown colors

The experimental markdown colors shouldn't be used for any arbitrary purpose outside of `x/markdown` package.  This commit unexports several types and functions, leaving `x/markdown` functions as the main use cases.

Additionally, there are a few additional changes to improve understanding and reduce the code involved:

1. Use of "ANSIColorCode" has been replaced in favor of "glamourStyleColor" to explain their purpose
2. Replace "ANSIColorCode.String()" and "ANSIColor.StrPtr()" for "glamourStyleColor.Code()" to simplify and clarify use over types
3. Add comments to explain importance of enumeration order
4. Expand test verifying color codes

andyfeller committed:
Unexport glamour color code function

This isn't really needed beyond this package, so unexporting it to avoid unintended use cases.

williammartin commented on pkg/x/markdown/accessibility_test.go:
```diff
@@ -0,0 +1,165 @@
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```
**Suggestion**

Mark the subtests with `t.Parallel()` since there's nothing that prevents them being parallelised. Same for all the other tests.

andyfeller replied:
TIL `t.Parallel()` cannot be used with `t.Setenv()`

That said, I got as many as I could.

williammartin reviewed (approved):
### `./bin/gh repo view` with and without `GH_ACCESSIBLE_COLORS=1`

<img width="1920" alt="image showing markdown rendering" src="https://github.com/user-attachments/assets/e59db178-85fd-4e61-87d9-0e495b99d488" />

And on a white background with `GH_ACCESSIBLE_COLORS=1`

<img width="877" alt="image" src="https://github.com/user-attachments/assets/202b34a8-6aca-4d14-b34c-908ef81646f3" />

### With config setting

```
➜  cli git:(trunk) ✗ gh config set accessible_colors enabled
! warning: 'accessible_colors' is not a known configuration key
```

<img width="944" alt="image" src="https://github.com/user-attachments/assets/1661ca33-88c3-4881-a97b-797091e78395" />

williammartin commented:
@andyfeller There is text in the images above "Image: screenshot of gh pr status → https://user-images.githubusercontent.com/98482/84171218-327e7a80-aa40-11ea-8cd1-5177fc2d0e72.png" and I'm having trouble convincing myself that it is using ANSI 4-bit colors.

<img width="785" alt="image" src="https://github.com/user-attachments/assets/198d3945-3e3d-45b4-8e7e-83db84e38443" />

From using `sequin` it looks like the control code is:

```
 CSI 37m: Foreground color: White
 CSI 38;5;243m: Foreground color: 243
```

So this looks like a 256-bit color code. I think this is the relevant style bits are:
 * https://github.com/charmbracelet/glamour/blob/d1d5125cfab42203495f15b846e901d0c1f37aaa/styles/styles.go#L233
 * https://github.com/charmbracelet/glamour/blob/d1d5125cfab42203495f15b846e901d0c1f37aaa/styles/styles.go#L442

---

It also looks to me as if [LinkText](https://github.com/charmbracelet/glamour/blob/d1d5125cfab42203495f15b846e901d0c1f37aaa/styles/styles.go#L434) would not be 16-bit, but I could be mistaken.

Here's the green `LinkText` "hub" on my dark background:

<img width="506" alt="image" src="https://github.com/user-attachments/assets/10c2787f-2fdc-4dab-b85d-b799d24995b4" />

Indeed, when I run it through `sequin` I see:

```
 CSI 0m: Reset style
 CSI 38;5;35;1m: Foreground color: 35, Bold
Text more detailed explanation
```

Which matches the `DarkStyleConfig` [LinkTest](https://github.com/charmbracelet/glamour/blob/d1d5125cfab42203495f15b846e901d0c1f37aaa/styles/styles.go#L224-L226)

williammartin reviewed (changes requested):
Requesting changes based on ^

andyfeller committed:
Expand a11y testing, fix inaccessible colors

This commit is focused on expanded accessibility testing and addressing discoveries from them:

1. `TestAccessibleDarkStyleConfigIs4Bit` and `TestAccessibleLightStyleConfigIs4Bit` will walk the Glamour StyleConfig type and check all elements
2. `Test_RenderAccessible` has been expanded with anchor and image elements

In addition, there are several fixes to logic and tests:

1. `Test_RenderAccessible` Go codeblock was missing replacement value, resulting in inaccurate test input
2. Dark and light accessible style logic was missing link text, image text, and H6 styling
3. Handling the `StyleBlock` of codeblocks which is unclear how it interacts with Chroma

andyfeller commented:
## Accessible colors should also be used when `GLAMOUR_STYLE` is `dark` or `light`

### Problem

When determining how to render markdown, the code is concerned with 2 things:

1. Is the terminal background `dark` or `light`?  Or should `none` colors be used? _(`theme`)_ 

2. How does the user want to stylize markdown? _(`style`)_
   - Using a built-in [`charmbracelet/glamour` style](https://github.com/cli/go-gh/blob/7a6e5543a7f13a372dfafa88fac99981496b4991/pkg/term/env.go#L139-L148) _(`ascii`, `auto`, `dark`, `dracula`, `tokyo-night`, `light`, `notty`, `pink`)_
   - Using a path to a [custom `charmbracelet/glamour` style](https://github.com/charmbracelet/glamour/tree/master?tab=readme-ov-file#styles) defined as JSON

Currently, **accessible colors are only applied if `style` has not been set**, which can be confusing if the user sets `GLAMOUR_STYLE` to `dark` or `light`:

https://github.com/cli/go-gh/blob/bde34d5cca49d621bce438a47b3ab94c4246856a/pkg/markdown/markdown.go#L33-L53

- `GH_ACCESSIBLE_COLORS=1 bin/gh repo view andyfeller/markdown-lorem-ipsum`

  As we see, `gh` is using the custom accessible themes when `GLAMOUR_STYLE` is unset:

  ![Screenshot of new accessible experience without glamour env variable](https://github.com/user-attachments/assets/edd743c6-0a72-4e9a-9fd6-67293c3cdee9)

- `GLAMOUR_STYLE=dark GH_ACCESSIBLE_COLORS=1 bin/gh repo view andyfeller/markdown-lorem-ipsum`

  However we see `gh` will fallback to the current markdown rendering experience even when the `style` and `theme` should be the same.

  ![Screenshot of new accessible experience with glamour env variable](https://github.com/user-attachments/assets/c4801c18-ea45-4043-807a-05eb52d4b4b5)

### Suggestion

This was honestly a bit confusing for me and I've been working on these changes for a while as I expected `GLAMOUR_STYLE=dark` or `GLAMOUR_STYLE=light` to also pick up accessible colors.

If this logic was enhanced to allow those 2 `styles` to fall into the same logic that applies accessible colors, I think we'll have fewer people surprised.

andyfeller committed:
Update to glamour v0.9.0 with contribution

This branch was pinned to the commit SHA after our contribution was merged, however our friends at Charm have cut a new release, which we should use instead.

andyfeller commented:
## Should tables be wrapped or not? 🤔 

One of the improvements in `charmbracelet/glamour@v0.9.0` is [greater control over table wrapping / overflow](https://github.com/charmbracelet/lipgloss/releases/tag/v1.1.0), which `glamour` now [does for tables by default](https://github.com/charmbracelet/glamour/blob/f43b1ad9ef09b10a93837e07ae2c18b66bceac5c/ansi/table.go#L35-L61):

```go
	ctx.table.lipgloss = table.New().Width(width).Wrap(true)
```

**UPDATE: This PR has been updated with upstream changes to disable table content wrapping until we can understand its impact to screenreader accessibility.**

<details>
<summary>
  <b>Expand for full thoughts</b>
</summary>
<p>

I'm going to enhance https://github.com/andyfeller/markdown-lorem-ipsum to contain an example for discussion as I'm unsure how this affects accessible screenreader experience, which is outside the scope of the issue and PR here.

### Current `gh` markdown table rendering experience

<img width="1128" alt="Screenshot 2025-03-13 at 5 54 04 PM" src="https://github.com/user-attachments/assets/492dc5b5-3caa-476a-a074-9748b2a55d17" />

### New `gh` experience table rendering experience

<img width="1166" alt="Screenshot 2025-03-13 at 5 53 14 PM" src="https://github.com/user-attachments/assets/360ccee0-dd8c-4366-b8cc-1bef0a959e91" />
<img width="1153" alt="Screenshot 2025-03-13 at 5 53 38 PM" src="https://github.com/user-attachments/assets/7964dd70-b81e-42d1-86da-4dca4ecab962" />


cc: @williammartin 


</p>
</details>

andyfeller committed:
Update `glamour` and disable table wrapping

This commit updates the `charmbracelet/glamour` dependency to pick up an unreleased change around markdown table rendering and table wrapping.

The concern previously was that table wrapping would cause issues in the current `gh` experience as well as make it difficult for screen readers to understand content.

andyfeller committed:
go mod tidy

andyfeller committed:
Parallelize tests

andyfeller commented:
@williammartin : I'm going to spend some time later tonight putting together a demo of the changes in the PR just so we can confirm the experience meets expectations beyond the tests and code review.  Though the fix for table wrapping hasn't been released, I think it's okay we pin to the commit for now to get this merged.

So re-re-requesting review and I'll follow up on demoing it later tonight.

**UPDATE: See PR body above for updated screenshots**

andyfeller commented:
Want to give a special shout out to our friends at [`charmbracelet`](https://github.com/charmbracelet/) who helped make this work possible with work on `glamour` and `lipgloss` in no specific order:

- @andreynering
- @aymanbagabas
- @bashbunni
- @caarlos0
- @meowgorithm

andyfeller committed:
Disable table wrapping for all

williammartin commented on go.mod:
```diff
@@ -1,27 +1,31 @@
-go 1.21
+go 1.23.0
+
+toolchain go1.23.7
```
Is this really required? Would think it is unusual to have a toolchain entry for a library.

andyfeller replied:
Great question; `go mod` did this in upgrading dependencies 🤦

andyfeller replied:
Removing for now

williammartin reviewed (approved):
Approved with one non blocking comment! Well done on all this, I know it was a lot.

andyfeller committed:
Remove toolchain directive from library

Discussing this with @williammartin, neither of us could reason why the `toolchain` directive was necessary for a library and choosing to remove.
